### PR TITLE
Sam Adapters for RxJava 2.x

### PR DESCRIPTION
**Pull request by thomasnield**
This is the SAM adapter implementation WIP for RxKotlin 2.x, initiated from the following discussions by @chris-horner:

https://github.com/ReactiveX/RxKotlin/issues/103
https://github.com/ReactiveX/RxKotlin/pull/109

### CODE REVIEW

**Comment by thomasnield** [comment]
I will look into Travis build errors later. For now, I see we need to create SAM adapters for the following operators, for `ObservableKt`, `FlowableKt`, `MaybeKt`, and `SingleKt` where applicable: 

* `ObservableKt.combineLatest()`
* `ObservableKt.zip()`
* `Observable#zipWith()`
* `Observable#withLatestFrom()`

There might be some needed for `flatMap()` and `concatMap()` overloads as well. Please let me know if I am missing any others.

**Comment by thomasnield** [comment]
Not quite sure why the Kotlin compiler is not liking `ObservableKt`, `FlowableKt`, `SingleKt`, and `MaybeKt`. It says these are duplicate JVM classes. Looking through the `.class` files and I'm not quite sure if I see any duplicates. I really like those names and want them to work. I'd rather change what they are clashing with. 

image

**Comment by thomasnield** [comment]
Renamed `ObservableKt` to `Observables`, `FlowableKt` to `Flowables`, etc... to resolve build issues at the moment.

File: src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import io.reactivex.functions.BiFunction
+import io.reactivex.functions.Function3
+import io.reactivex.functions.Function4
+import io.reactivex.functions.Function5
+import org.reactivestreams.Publisher
+
+
+object Flowables {
+
+    inline fun <T1,T2,R> combineLatest(source1: Flowable<T1>, source2: Flowable<T2>, crossinline combineFunction: (T1, T2) -> R) =
```
Comment (chris-horner) [inline src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt:10]: nit: I know people aren't often going to be reading this file, but I feel `combiner` is a nicer name than `combineFunction`.

Comment (thomasnield) [reply src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt:10]: Okay Ill change that.

File: src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+object Flowables {
+
+    inline fun <T1,T2,R> combineLatest(source1: Flowable<T1>, source2: Flowable<T2>, crossinline combiner: (T1, T2) -> R) =
+            Flowable.combineLatest(source1, source2,
+                    BiFunction<T1, T2, R> { t1, t2 -> combiner(t1,t2) })!!
+
+    inline fun <T1,T2,T3,R> combineLatest(source1: Flowable<T1>, source2: Flowable<T2>, source3: Flowable<T3>, crossinline combiner: (T1,T2, T3) -> R) =
+            Flowable.combineLatest(source1, source2,source3,
+                    io.reactivex.functions.Function3<T1, T2, T3, R> { t1: T1, t2: T2, t3: T3 -> combiner(t1,t2, t3) })!!
```
Comment (chris-horner) [inline src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt]: Any particular reason for the fully qualified `Function`s?

**Comment by thomasnield** [comment]
My compiler struggled if I didn't.

**Comment by chris-horner** [comment]
Have a look at the methods I've defined that you've copied across. `Function3`, `4, and `5` all work without being fully qualified. Perhaps IntelliJ's not grabbing the correct import for you?

**Comment by thomasnield** [comment]
I'll follow up this weekend on this.

**Comment by thomasnield** [comment]
@chris-horner Couldn't figure out the `BiFunction` not inferring correctly, but was able to clean up the others.

File: build.gradle [hunk 3]
```diff
@@ -1,5 +1,5 @@
 buildscript {
-    ext.kotlin_version = '1.1.0'
+    ext.kotlin_version = '1.1.2'
```
Comment (stepango) [inline build.gradle:2]: 1.1.2-2 is the latest one

File: src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt [hunk 4]
```diff
@@ -0,0 +1,144 @@
+package io.reactivex.rxkotlin
+
+import io.reactivex.Flowable
+import io.reactivex.functions.*
```
Comment (stepango) [inline src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt:4]: avoid wildcard imports

File: src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt [hunk 5]
```diff
@@ -0,0 +1,144 @@
+package io.reactivex.rxkotlin
```
Comment (stepango) [inline src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt:1]: code formatting in all file

File: src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+import io.reactivex.Flowable
+import io.reactivex.functions.*
+import org.reactivestreams.Publisher
+
+
+object Flowables {
+
+    inline fun <T1,T2,R> combineLatest(source1: Flowable<T1>, source2: Flowable<T2>, crossinline combineFunction: (T1, T2) -> R) =
+            Flowable.combineLatest(source1, source2,
+                    BiFunction<T1, T2, R> { t1, t2 -> combineFunction(t1,t2) })!!
```
Comment (stepango) [inline src/main/kotlin/io/reactivex/rxkotlin/Flowables.kt:12]: why `!!`?

**Review by stepango — CHANGES_REQUESTED** [review]

**Pull request merged**